schedule/scheduler.go

```diff
@@ -14,159 +18,301 @@ import (
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		// A little jitter between scheduledTasks to reduce thundering herd effects on api.
+		wait.SleepWithJitter(s.jobDelay, 2)
+	}
+
+	// Wait until the context is done and handle new jobs as they are submitted.
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case newTask := <-s.jobQueue:
+			newTask.log = log.With("id", newTask.id, "name", newTask.Name)
+			s.jobs[newTask.id] = newTask
+			newTask.log.Infow("new job received")
+			go s.execute(newTask, s.workerJobComplete)
+		case <-s.scheduledJobComplete:
+			// A job has completed
+			s.scheduledJobsRunning--
+			if s.scheduledJobsRunning == 0 {
+				log.Info("no scheduled jobs running")
+			}
+		case <-s.workerJobComplete:
+			s.workerJobsRunning--
+			if s.workerJobsRunning == 0 {
+				log.Info("no worker jobs running")
 			}
+		}
+	}
+}
+
+func (s *Scheduler) Start(id JobID) error {
+	job, ok := s.jobs[id]
+	if !ok {
+		return xerrors.Errorf("starting worker ID: %d not found", id)
+	}
 
-			// Keep this task running forever
-			doneFirstRun := false
-			for {
+	if job.running {
+		return xerrors.Errorf("starting worker ID: %d already running", id)
+	}
+	job.log.Info("starting job")
+	go s.execute(job, s.workerJobComplete)
+	return nil
+}
 
-				// Is the context done?
-				select {
-				case <-ctx.Done():
-					return
-				default:
-				}
+func (s *Scheduler) Stop(id JobID) error {
+	job, ok := s.jobs[id]
+	if !ok {
+		return xerrors.Errorf("starting worker ID: %d not found", id)
+	}
 
-				if doneFirstRun {
-					log.Infow("restarting task", "task", tc.Name, "delay", tc.RestartDelay)
-					if tc.RestartDelay > 0 {
-						time.Sleep(tc.RestartDelay)
-					}
-				} else {
-					log.Infow("running task", "task", tc.Name)
-					doneFirstRun = true
-				}
+	if !job.running {
+		return xerrors.Errorf("starting worker ID: %d already running", id)
+	}
+	job.log.Info("stopping job")
+	if job.cancel != nil {
+		job.cancel()
+	}
+	job.running = false
+	return nil
+}
+
+type JobResult struct {
+	ID                  JobID
+	Running             bool
+	Name                string
+	RestartOnFailure    bool
+	RestartOnCompletion bool
+	RestartDelay        time.Duration
+}
+
+var InvalidJobID = JobID(0)
+
+type JobID int
+
+func (s *Scheduler) Jobs() []JobResult {
+	if len(s.jobs) == 0 {
+		return nil
+	}
+	var out []JobResult
+	for _, j := range s.jobs {
+		out = append(out, JobResult{
+			ID:                  j.id,
+			Running:             j.running,
+			Name:                j.Name,
+			RestartOnFailure:    j.RestartOnFailure,
+			RestartOnCompletion: j.RestartOnCompletion,
+			RestartDelay:        j.RestartDelay,
+		})
+	}
+	return out
+}
+
+func (s *Scheduler) execute(tc *JobConfig, complete chan struct{}) {
+	ctx, cancel := context.WithCancel(s.context)
+	tc.cancel = cancel
+	tc.running = true
+	// Report job is complete when this goroutine exits
+	defer func() {
+		complete <- struct{}{}
+		tc.running = false
+		tc.cancel()
```

Comment:
cancel can be nil? You checked for it earlier. 

Probably best to set to nil once the job has completed.

Reply:
It can never be `nil`, all jobs are stated via the execute method and a cancelFunc is always assigned to a job at the beginning of the method. I've dropped the error check as it's not needed. I can't come up with a strong reason for setting cancel to nil, what do you think?